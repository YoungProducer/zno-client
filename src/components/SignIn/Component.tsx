/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 February 2020
 *
 * Component which allows user to make sign in
 */

/** External imports */
import React, { useCallback, useState, useEffect } from 'react';
import classNames from 'classnames';
import Paper from '@material-ui/core/Paper';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import CheckBox from '@material-ui/core/Checkbox';
import InputBase from '@material-ui/core/InputBase';
import Typography from '@material-ui/core/Typography';
import EmailIcon from '@material-ui/icons/Email';
import LockIcon from '@material-ui/icons/Lock';
import { makeStyles, createStyles, Theme, withStyles, WithStyles } from '@material-ui/core/styles';
import { useHistory } from 'react-router-dom';

/** Application's imports */
import { TSignInProps } from './container';
import NavigationLink from 'components/NavigationLink';
import Input from 'components/Input';
import Logo from 'img/logo';
import Check from 'public/images/check.svg';

/** Define classes as hook */
const useStyles = makeStyles((theme: Theme) => createStyles({
    backdrop: {
        background: '#fff',
        width: '100%',
        height: '100vh',
    },
    container: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
        height: '100%',
    },
    root: {
        width: 'min-content',
        borderRadius: 4,
        filter: `drop-shadow(2.828px 2.828px 3.5px rgba(227,207,207,0.9))`,
        backgroundColor: '#ffffff',
        textAlign: 'center',
    },
    innerContainer: {
        borderRadius: 4,
        backgroundColor: '#f4f5f7',
        padding: theme.spacing(4),
    },
    logoContainer: {
        paddingTop: theme.spacing(1),
        paddingBottom: theme.spacing(1),
    },
    textField: {
        width: 230,
    },
    fields: {
        paddingBottom: theme.spacing(3),
    },
    actions: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    block: {
        width: 'max-content',
    },
    buttonBlock: {
        width: '100%',
        marginTop: theme.spacing(2),
    },
    button: {
        filter: `drop-shadow(0px 3px 2.5px rgba(0,0,0,0.16))`,
    },
    link: {
        color: theme.palette.primary.main,
    },
    icon: {
        color: '#555',
    },
    checkBoxIcon: {
        marginLeft: 2,
        borderRadius: 3,
        width: 20,
        height: 20,
        filter: `drop-shadow(2.828px 2.828px 3.5px rgba(227,207,207,0.9))`,
        backgroundColor: '#fff',
        backgroundImage: 'linear-gradient(180deg,hsla(0,0%,100%,.8),hsla(0,0%,100%,0))',
        '$root.Mui-focusVisible &': {
            outline: '2px auto rgba(19,124,189,.6)',
            outlineOffset: 2,
        },
        'input:hover ~ &': {
            backgroundColor: '#ebf1f5',
        },
        'input:disabled ~ &': {
            boxShadow: 'none',
            background: 'rgba(206,217,224,.5)',
        },
    },
    checkBoxСheckedIcon: {
        backgroundImage: 'linear-gradient(180deg,hsla(0,0%,100%,.8),hsla(0,0%,100%,0))',
        position: 'relative',
        '&:before': {
            position: 'absolute',
            top: 2,
            left: 2,
            backgroundImage: `url(${Check})`,
            width: 20,
            height: 20,
            backgroundSize: '16px, 16px',
            display: 'block',
            content: '""',
            backgroundRepeat: 'no-repeat',
        },
        'input:hover ~ &': {
            // backgroundColor: '#106ba3',
        },
    },
    checkBox: {
        width: 400,
    },
}));

/**
 * Create hook to make component stateless.
 * Compute all values from text fields.
 * Returns fields data and button onClick property.
 */
const useSignInElements = (props: TSignInProps) => {
    const {
        isLoggedIn,
        errorFields,
        fieldsMessages,
        fetchSignIn,
        setSignInErrorFieldsToDefault,
        setSignInFieldsMessagesToDefault,
    } = props;

    const history = useHistory();

    useEffect(() => {
        if (isLoggedIn) {
            const next = history.location.state
                ? `${(history.location.state as any).from.pathname}${(history.location.state as any).from.search}`
                : undefined;
            // const next: string = new URLSearchParams(history.location.state.).get('next');
            history.push(next || '/');
        }
    }, [isLoggedIn]);

    const clearFields = () => {
        setSignInErrorFieldsToDefault();
        setSignInFieldsMessagesToDefault();
    };

    /** User email */
    const [email, setEmail] = useState<string>('');
    const handleChangeEmail =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setEmail(event.target.value);
            if (errorFields.email) {
                clearFields();
            }
        }, [errorFields.email]);

    /** User password */
    const [password, setPassword] = useState<string>('');
    const handleChangePassword =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setPassword(event.target.value);
            if (errorFields.password) {
                clearFields();
            }
        }, [errorFields.password]);

    /** SignUp button handle */
    const handleFetchSignIn = useCallback(() => fetchSignIn({
        email,
        password,
    }), [email, password]);

    return {
        emailField: {
            onChange: handleChangeEmail,
            value: email,
            error: errorFields.email,
            helperText: fieldsMessages.email,
        },
        passwordField: {
            onChange: handleChangePassword,
            value: password,
            error: errorFields.password,
            helperText: fieldsMessages.password,
        },
        signInButton: {
            onClick: handleFetchSignIn,
        },
    };
};

/** Create component */
const Component = (props: TSignInProps) => {
    /** Create styles */
    const classes = useStyles({});

    /** Get fields data from hook */
    const { emailField, passwordField, signInButton } = useSignInElements(props);

    return (
        <div
            className='auth-background'
        >
            <Container maxWidth='lg' className={classes.container}>
                <Paper
                    className={classes.root}
                    elevation={0}
                    square
                >
                    <div className={classes.logoContainer}>
                        <Logo />
                    </div>
                    <div className={classes.innerContainer}>
                        <Grid
                            container
                            direction='column'
                            alignItems='center'
                            justify='flex-start'
                            item
                            spacing={3}
                            className={classes.block}
                        >
                            <Grid item>
                                <Input
                                    color='primary'
                                    type='email'
                                    placeholder='Емеїл'
                                    startAdornment={<EmailIcon className={classes.icon}/>}
                                    data-testid='signin-email-input'
                                    {...emailField}
                                />
                            </Grid>
                            <Grid item>
                                <Input
                                    color='primary'
                                    type='password'
                                    placeholder='Пароль'
                                    startAdornment={<LockIcon className={classes.icon}/>}
                                    data-testid='signin-password-input'
                                    {...passwordField}
                                />
                            </Grid>
                            <Grid item>
                                <FormControl component='fieldset' className={classes.checkBox}>
                                    <FormControlLabel
                                        control={
                                            <CheckBox
                                                disableRipple
                                                disableTouchRipple
                                                disableFocusRipple
                                                color='primary'
                                                checkedIcon={<span className={classNames(classes.checkBoxIcon, classes.checkBoxСheckedIcon)} />}
                                                icon={<span className={classes.checkBoxIcon}/>}
                                            />
                                        }
                                        label={`Запам'ятати мене`}
                                    />
                                </FormControl>
                            </Grid>
                            <Grid item>
                                <Button
                                    color='primary'
                                    variant='contained'
                                    disableElevation
                                    className={classes.button}
                                    data-testid='signin-button'
                                    { ...signInButton }
                                >
                                    Увійти
                                </Button>
                            </Grid>
                            <Grid item>
                                <Typography align='center'>
                                    Не зареєстровані?
                                    <NavigationLink
                                        navLink={{
                                            to: '/auth/signup',
                                        }}
                                        className={classes.link}
                                    >
                                        Зареєструватися
                                    </NavigationLink>
                                </Typography>
                            </Grid>
                        </Grid>
                    </div>
                </Paper>
            </Container>
        </div>
    );
};

export default Component;
