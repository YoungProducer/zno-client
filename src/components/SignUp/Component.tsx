/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 25 February 2020
 *
 * Component which allows create user's account.
 */

// External imports
import React, { useState, useCallback, useEffect, memo, useMemo } from 'react';
import Paper from '@material-ui/core/Paper';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';

// Application's imports
import { TSignUpProps } from './container';
import NavigationLink from 'components/NavigationLink';

// Define classes as hook
const useStyles = makeStyles((theme: Theme) => createStyles({
    backdrop: {
        background: '#fff',
        width: '100%',
        height: '100vh',
    },
    root: {
        position: 'fixed',
        padding: 15,
        left: '50%',
        top: '50%',
        transform: `translate(-50%, -50%)`,
        width: 'min-content',
    },
    textField: {
        width: 230,
    },
    fields: {
        paddingBottom: theme.spacing(3),
    },
    actions: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    block: {
        width: 'max-content',
    },
    button: {
        width: 230,
    },
    link: {
        color: theme.palette.primary.main,
    },
}));

/**
 * Create hook to make component stateless.
 * Compute all values from text fields.
 * Returns fields data and button onClick property.
 */
const useSignUpElements = (props: TSignUpProps) => {
    const {
        errorFields,
        fieldsMessages,
        fetchSignUp,
        setSignUpErrorFieldsToDefault,
        setSignUpFieldsMessagesToDefault,
    } = props;

    const clearFields = () => {
        setSignUpErrorFieldsToDefault();
        setSignUpFieldsMessagesToDefault();
    };

    /** User email */
    const [email, setEmail] = useState<string>('');
    const handleChangeEmail =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setEmail(event.target.value);
            if (errorFields.email) {
                clearFields();
            }
        }, [errorFields.email]);

    /** User password */
    const [password, setPassword] = useState<string>('');
    const handleChangePassword =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setPassword(event.target.value);
            if (errorFields.password) {
                clearFields();
            }
        }, [errorFields.password]);

    /** Confirmation password, should be the same to the password */
    const [confPassword, setConfPassword] = useState<string>('');
    const handleChangeConfPassword =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setConfPassword(event.target.value);
            if (errorFields.confPassword) {
                clearFields();
            }
        }, [errorFields.confPassword]);

    /** SignUp button handle */
    const handleFetchSignUp = useCallback(() => fetchSignUp({
        email,
        password,
        confPassword,
    }), [email, password, confPassword]);

    return {
        emailField: {
            onChange: handleChangeEmail,
            value: email,
            error: errorFields.email,
            helperText: fieldsMessages.email,
        },
        passwordField: {
            onChange: handleChangePassword,
            value: password,
            error: errorFields.password,
            helperText: fieldsMessages.password,
        },
        confPasswordField: {
            onChange: handleChangeConfPassword,
            value: confPassword,
            error: errorFields.confPassword,
            helperText: fieldsMessages.confPassword,
        },
        signUpButton: {
            onClick: handleFetchSignUp,
        },
    };
};

const Component = (props: TSignUpProps) => {
    /** Create object with classes to use them */
    const classes = useStyles({});

    /** Get fields data from hook */
    const {
        emailField,
        passwordField,
        confPasswordField,
        signUpButton,
    } = useSignUpElements(props);

    return (
        <div className={classes.backdrop}>
            <Paper
                className={classes.root}
                elevation={2}
                square
            >
                <Typography
                    color='primary'
                    variant='h6'
                    align='center'
                >
                    Реєстрація
                </Typography>
                <Grid
                    container
                    direction='row'
                    spacing={2}
                    className={classes.fields}
                >
                    <Grid
                        container
                        direction='row'
                        alignItems='center'
                        item
                        spacing={3}
                        className={classes.block}
                    >
                        <Grid item>
                            <TextField
                                className={classes.textField}
                                color='primary'
                                label='Емеїл'
                                type='email'
                                { ...emailField }
                            />
                        </Grid>
                        <Grid item>
                            <TextField
                                className={classes.textField}
                                color='primary'
                                label='Пароль'
                                type='password'
                                { ...passwordField }
                            />
                        </Grid>
                    </Grid>
                    <Grid
                        container
                        direction='row'
                        alignItems='center'
                        item
                        spacing={3}
                        className={classes.block}
                    >
                        <Grid item>
                            <TextField
                                className={classes.textField}
                                color='primary'
                                label='Пароль підтвердження'
                                type='password'
                                { ...confPasswordField }
                                { ...{ 'data-testid': 'conf-password' } }
                            />
                        </Grid>
                        <Grid item>
                            <Button
                                color='primary'
                                variant='contained'
                                disableElevation
                                className={classes.button}
                                data-testid='signup-button'
                                { ...signUpButton }
                            >
                                Зареєструватися
                            </Button>
                        </Grid>
                    </Grid>
                </Grid>
                <Typography align='center'>
                    Вже зареєстровані?
                    <NavigationLink
                        navLink={{
                            to: '/auth/signin',
                        }}
                        className={classes.link}
                    >
                        Увійти
                    </NavigationLink>
                </Typography>
            </Paper>
        </div>
    );
};

export default Component;
