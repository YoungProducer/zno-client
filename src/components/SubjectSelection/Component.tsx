/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Component to select subjects.
 */

/** External imports */
import React from 'react';
import Container from '@material-ui/core/Container';
import CSSBaseLine from '@material-ui/core/CssBaseline';
import Grid from '@material-ui/core/Grid';

/** Application's imports */
import { TSubjectSelectionProps } from './container';
import SubjectTile from './SubjectTile';

/** Create component */
const Component = (props: TSubjectSelectionProps) => {
    const {
        classes,
        subjectsList,
        fetchSubjectsNames,
    } = props;

    return (
        <div className={classes.root}>
            <Grid container spacing={1}>
                {subjectsList.map((subject, index) => (
                    <SubjectTile subject={subject} key={index} />
                ))}
            </Grid>
        </div>
    );
};

export default Component;
