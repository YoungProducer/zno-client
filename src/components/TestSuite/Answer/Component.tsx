/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 23 March 2020
 */

/** External imports */
import React, { useState } from 'react';
import classNames from 'classnames';
import ButtonBase from '@material-ui/core/ButtonBase';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

/** Application's imports */
import { TAnswerProps } from './container';
import Single from './Single';
import Relations from './Relations';
import { TAnswerType } from 'api';

export const markUp = ['A', 'Б', 'В', 'Г', 'Д'];

interface IAnswerProps {
    type: TAnswerType;
    taskIndex: number;
}

const Component = ({ type, taskIndex }: IAnswerProps) => {
    return (
        <>
            { type === 'SINGLE' && <Single taskIndex={taskIndex}/>}
            { type === 'RELATIONS' && <Relations taskIndex={taskIndex}/>}
        </>
    );
};

export default Component;
