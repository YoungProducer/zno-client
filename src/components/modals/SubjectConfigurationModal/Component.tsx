/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 4 March 2020
 *
 * Component which allows to configure test options
 * and allows to select which test you want to complete.
 */

/** External imports */
import React, { useState, useCallback, useMemo, useEffect } from 'react';
import Dialog from '@material-ui/core/Dialog';
import MuiDialogTitle from '@material-ui/core/DialogTitle';
import MuiDialogContent from '@material-ui/core/DialogContent';
import MuiDialogActions from '@material-ui/core/DialogActions';
import Typography from '@material-ui/core/Typography';
import MenuItem from '@material-ui/core/MenuItem';
import TextField from '@material-ui/core/TextField';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormControl from '@material-ui/core/FormControl';
import FormLabel from '@material-ui/core/FormLabel';
import Button from '@material-ui/core/Button';
import { createStyles, Theme, withStyles, WithStyles } from '@material-ui/core/styles';

/** Application's imports */
import {
    ETestTypes,
    EExamTypes,
    TSubjectConfigurationModalProps,
} from './container';

/** Define Material UI styles as hook */

const styles = (theme: Theme) =>
    createStyles({
        root: {
            margin: 0,
            padding: theme.spacing(2),
        },
    });

/** Create components wrapped by withStyles */
export interface DialogTitleProps extends WithStyles<typeof styles> {
    id: string;
    children: React.ReactNode;
}

const DialogTitle = withStyles(styles)((props: DialogTitleProps) => {
    const { children, classes, ...other } = props;
    return (
        <MuiDialogTitle disableTypography className={classes.root} {...other}>
            <Typography variant="h6">{children}</Typography>
        </MuiDialogTitle>
    );
});

const DialogContent = withStyles((theme: Theme) => ({
    root: {
        padding: theme.spacing(2),
        paddingLeft: theme.spacing(3),
        paddingRight: theme.spacing(3),
    },
}))(MuiDialogContent);

/**
 * Define custom hook to save state of configuration.
 * This hook returns object with settings
 * for all input elements and dialog.
 */
const useSubjectConfigurationElements = (props: TSubjectConfigurationModalProps) => {
    /** Destruct props */
    const {
        dialogVisible,
        subjectThemes,
        subjectExams,
        subSubjectsNames,
        subSubjectsThemes,
        toggleSubjectConfigurationDialog,
    } = props;

    /** Responsible for themes or exams selection */
    const [testType, setTestType] = useState<ETestTypes>('' as ETestTypes);

    /** Responsible for theme selection of specific sub-subject */
    const [subSubject, setSubSubject] = useState<string>(subSubjectsNames !== null ? subSubjectsNames[0] : '');

    /** Responsible for theme selection */
    const [theme, setTheme] = useState<string>(subjectThemes !== null ? subjectThemes[0] : '');

    /** Responsible for exam type selection */
    const [examType, setExamType] = useState<EExamTypes>('' as EExamTypes);

    /** Respobsible for exam selection */
    const [exam, setExam] = useState<string>('');

    /**
     * Handle onChange event of test-type field.
     * If value equals THEMES
     * then update property 'theme'.
     */
    const handleChangeTestType = useCallback(
        (event: React.ChangeEvent<HTMLInputElement>) => {
            setTestType(event.target.value as ETestTypes);

            if (event.target.value === ETestTypes.THEMES) {
                if (subjectThemes !== null) {
                    setTheme(subjectThemes[0]);
                }

                if (subSubject !== '' && subSubjectsThemes !== null) {
                    setTheme(subSubjectsThemes[subSubject] ? subSubjectsThemes[subSubject][0] : '');
                }
            }
        }, [subjectThemes, subSubject, subSubjectsThemes]);

    /** Handle onChange event of sub-subject text-field */
    const handleChangeSubSubject = useCallback(
        (event: React.ChangeEvent<HTMLInputElement>) => {
            const value = event.target.value;

            setSubSubject(value);

            if (testType === ETestTypes.THEMES && subSubjectsThemes !== null) {
                setTheme(subSubjectsThemes[value] ? subSubjectsThemes[value][0] : '');
            } else setTheme('');
        }, [subSubjectsThemes, testType]);

    /** Handle onChange evenet of theme text-field */
    const handleChangeTheme = useCallback(
        (event: React.ChangeEvent<HTMLInputElement>) =>
            setTheme(event.target.value),
        []);

    /**
     * Handle onChange event of exam-type field.
     * Update exam value if 'testType' equals EXAMS.
     */
    const handleChangeExamType = useCallback(
        (event: React.ChangeEvent<HTMLInputElement>) => {
            setExamType(event.target.value as EExamTypes);

            if (testType === ETestTypes.EXAMS && subjectExams !== null) {
                if (event.target.value === EExamTypes.SESSIONS) {
                    setExam(subjectExams.sessions !== null ? subjectExams.sessions[0] : '');
                }

                if (event.target.value === EExamTypes.TRAININGS) {
                    setExam(subjectExams.trainings !== null ? subjectExams.trainings[0] : '');
                }
            }
        }, [subjectExams, testType]);

    /** Handle onChange event of exam text field */
    const handleChangeExam = useCallback(
        (event: React.ChangeEvent<HTMLInputElement>) =>
            setExam(event.target.value),
        []);

    /** Memoized value which returns list of sub-subjects */
    const subSubjects = useMemo(() => {
        if (subSubjectsNames === null || testType !== ETestTypes.THEMES) return null;

        return subSubjectsNames.map(subject => (
            <MenuItem key={subject} value={subject}>
                {subject}
            </MenuItem>
        ));
    }, [subSubjectsNames, testType]);

    /** Responsible for displaying of sub-subject selection */
    const displaySubSubjectSelection = useMemo(() =>
        subSubjectsNames !== null && testType === ETestTypes.THEMES,
    [testType, subSubjectsNames]);

    /** Memoized value which returns list of themes */
    const themes = useMemo(() => {
        /**
         * If subSubjectsThemes is not null and some subSubject is selected
         * and testType equals 'THEMES'
         * then this function will returns themes list for selected subSubject.
         */
        if (subSubjectsThemes !== null && subSubject !== '' && testType === ETestTypes.THEMES) {
            return subSubjectsThemes[subSubject] ? subSubjectsThemes[subSubject].map(theme => (
                <MenuItem key={theme} value={theme}>
                    {theme}
                </MenuItem>
            )) : null;
        }

        /**
         * If subjectThemes is not null and subSubjectsThemes is null
         * and testType equals 'THEMES'
         * then this function will returns themes list for current subject.
         */
        if (subjectThemes !== null && subSubjectsThemes === null && testType === ETestTypes.THEMES) {
            return subjectThemes.map(theme => (
                <MenuItem key={theme} value={theme}>
                    {theme}
                </MenuItem>
            ));
        }

        return null;
    }, [subjectThemes, testType, subSubject, subSubjectsThemes]);

    /** Responsible for displaying of theme selection */
    const displayThemeSelection = useMemo(() =>
        (subSubjectsThemes !== null || subjectThemes !== null) && testType === ETestTypes.THEMES,
    [testType, subjectThemes, subSubjectsThemes]);

    /**
     * Responsible for displaying radio group
     * which allows to select which type of exams you want to choose
     * 'sessions' or 'trainings'
     */
    const displayExamTypeSelection = useMemo(() =>
        testType === ETestTypes.EXAMS && subjectExams !== null,
    [testType, subjectExams]);

    /**
     * Memoized value which returns list of exams
     * from sessions or traingings
     */
    const exams = useMemo(() => {
        /** Check is 'testType' equals 'EXAMS' and 'subjectExams' is not null */
        if (testType === ETestTypes.EXAMS && subjectExams !== null) {
            /**
             * If 'examType' equals 'TRAININGS'
             * and 'trainings' is not null
             * then return list trainings exams.
             */
            if (examType === EExamTypes.TRAININGS && subjectExams.trainings !== null) {
                return subjectExams.trainings.map(exam => (
                    <MenuItem key={exam} value={exam}>
                        {exam}
                    </MenuItem>
                ));
            }

            /**
             * If 'examType' equals 'SESSIONS'
             * and 'sessions' is not null
             * then return list sessions exams.
             */
            if (examType === EExamTypes.SESSIONS && subjectExams.sessions !== null) {
                return subjectExams.sessions.map(exam => (
                    <MenuItem key={exam} value={exam}>
                        {exam}
                    </MenuItem>
                ));
            }
        }

        return null;
    }, [subjectExams, testType, examType]);

    return {
        displayThemeSelection,
        displaySubSubjectSelection,
        displayExamTypeSelection,
        subSubjects,
        themes,
        exams,
        dialog: {
            open: dialogVisible,
            onClose: () => toggleSubjectConfigurationDialog(false),
        },
        selectTypeField: {
            value: testType,
            onChange: handleChangeTestType,
        },
        selectSubSubjectField: {
            value: subSubject,
            onChange: handleChangeSubSubject,
        },
        selectThemeField: {
            value: theme,
            onChange: handleChangeTheme,
        },
        selectExamTypeField: {
            value: examType,
            onChange: handleChangeExamType,
        },
        selectExamField: {
            value: exam,
            onChange: handleChangeExam,
        },
    };
};

/** Create component */
const Component = (props: TSubjectConfigurationModalProps) => {
    /** Get elements from custom hook */
    const {
        dialog,
        themes,
        subSubjects,
        exams,
        selectTypeField,
        selectThemeField,
        selectSubSubjectField,
        selectExamTypeField,
        selectExamField,
        displayThemeSelection,
        displaySubSubjectSelection,
        displayExamTypeSelection,
    } = useSubjectConfigurationElements(props);

    return (
        <Dialog
            {...dialog}
            aria-labelledby='subject-config-title'
            aria-describedby=''
        >
            <DialogTitle id='subject-config-title'>
                Налаштування тесту
            </DialogTitle>
            <DialogContent>
                <FormControl component='fieldset'>
                    <FormLabel component='legend'>Оберіть тип тесту</FormLabel>
                    <RadioGroup
                        aria-label='select-test-type'
                        name='select-test-type'
                        data-testid='select-test-type'
                        {...selectTypeField}
                    >
                        <FormControlLabel
                            value={ETestTypes.THEMES}
                            control={<Radio color='primary' />}
                            label={ETestTypes.THEMES}
                            labelPlacement='end'
                        />
                        <FormControlLabel
                            value={ETestTypes.EXAMS}
                            control={<Radio color='primary' />}
                            label={ETestTypes.EXAMS}
                            labelPlacement='end'
                        />
                    </RadioGroup>
                </FormControl>

                { displaySubSubjectSelection && (
                    <FormControl component='div'>
                        <FormLabel component='legend'>Оберіть предмет</FormLabel>
                        <TextField
                            id='standard-select-currency'
                            select={true}
                            {...selectSubSubjectField}
                            margin='none'
                            variant='outlined'
                            {...{ 'data-testid': 'select-sub-subject' }}
                        >
                            {subSubjects}
                        </TextField>
                    </FormControl>
                )}

                { displayThemeSelection && (
                    <FormControl component='div' data-testid='abc'>
                        <FormLabel component='legend'>Оберіть тему</FormLabel>
                        <TextField
                            id='standard-select-currency'
                            select={true}
                            {...selectThemeField}
                            margin='none'
                            variant='outlined'
                            {...{ 'data-testid': 'select-subject-theme' }}
                        >
                            {themes}
                        </TextField>
                    </FormControl>
                )}

                { displayExamTypeSelection && (
                    <>
                        <FormControl component='fieldset'>
                            <FormLabel component='legend'>Оберіть тип тесту</FormLabel>
                            <RadioGroup
                                aria-label='position'
                                name='position'
                                {...selectExamTypeField}
                                data-testid='select-exam-type'
                            >
                                <FormControlLabel
                                    value={EExamTypes.TRAININGS}
                                    control={<Radio color='primary' />}
                                    label={EExamTypes.TRAININGS}
                                    labelPlacement='end'
                                />
                                <FormControlLabel
                                    value={EExamTypes.SESSIONS}
                                    control={<Radio color='primary' />}
                                    label={EExamTypes.SESSIONS}
                                    labelPlacement='end'
                                />
                            </RadioGroup>
                        </FormControl>

                        <FormControl component='div'>
                            <FormLabel
                                component='legend'
                                data-testid='select-exam-title'
                            >
                                { selectExamTypeField.value === EExamTypes.TRAININGS
                                    ? 'Виберіть тренувальний варіант'
                                    : 'Виберіть варіант ЗНО'
                                }
                            </FormLabel>
                            <TextField
                                id='standard-select-currency'
                                select={true}
                                {...selectExamField}
                                margin='none'
                                variant='outlined'
                                {...{ 'data-testid': 'select-exam' }}
                            >
                                {exams}
                            </TextField>
                        </FormControl>
                    </>
                )}
            </DialogContent>
        </Dialog>
    );
};

/** Export component */
export default Component;
