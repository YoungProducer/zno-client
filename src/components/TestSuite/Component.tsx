/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 20 March 2020
 *
 * Component which download images for test suite
 * display answers and allow to pass different test suites.
 */

/** External imports */
import React, { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';
import Typography from '@material-ui/core/Typography';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

/** Application's imports */
import { TTestSuiteProps } from './container';
import { ITestSuiteCredentials } from 'api';
import Answer from './Answer';
import TaskSelection from './TaskSelection';

/** Define Material classes as hook */
const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            paddingTop: theme.spacing(4),
        },
        title: {
            fontSize: '1.4375rem',
            color: '#b19898',
        },
        img: {

        },
        answersBlock: {

        },
        tasksSelectionBlock: {
            marginTop: theme.spacing(3),
        },
    }));

const useSearch = () => {
    /** Define posible search params array */
    const searchNames = ['subjectId', 'subSubjectId', 'theme', 'session', 'training'];

    /** Get location */
    const location = useLocation();

    /** Get search */
    const search = location.search;

    /** Create search params instance */
    const searchParams = new URLSearchParams(search);

    /** Extract data from search */
    const searchData = searchNames.reduce((acc, curr) => {
        const param = searchParams.get(curr);

        if (param !== null) {
            return {
                ...acc,
                [curr]: param,
            };
        }

        return acc;
    }, {});

    return searchData as ITestSuiteCredentials;
};

const useInitTestSuite = (props: TTestSuiteProps) => {
    /** Extract props */
    const { fetchTestSuite } = props;

    /** Get search data */
    const searchData = useSearch();

    useEffect(() => {
        fetchTestSuite(searchData);
    }, []);
};

const useTestSuiteFileds = (props: TTestSuiteProps) => {
    const [currentTask, setCurrentTask] = useState<number>(0);

    return {
        task: {
            current: currentTask,
            set: setCurrentTask,
        },
    };
};

const Component = (props: TTestSuiteProps) => {
    /** Init classes */
    const classes = useStyles({});

    /** Destruct props */
    const { answers, name, tasksImages } = props;

    useInitTestSuite(props);

    const { task } = useTestSuiteFileds(props);

    return (
        <div className={classes.root}>
            <Typography className={classes.title}>{name}</Typography>
            <img
                className={classes.img}
                src={tasksImages[task.current]}
            />
            <div className={classes.answersBlock}>
                { answers.length !== 0 &&
                    <Answer
                        type={answers[task.current].type}
                        taskIndex={task.current}
                    />
                }
            </div>
            <div className={classes.tasksSelectionBlock}>
                <TaskSelection />
            </div>
        </div>
    );
};

export default Component;
