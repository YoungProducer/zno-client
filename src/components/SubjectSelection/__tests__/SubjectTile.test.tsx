/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Create tests suites for SubjectTile component.
 */

/** External imports */
import React from 'react';
import { shallow } from 'enzyme';

/** Application's imports */
import Component, { ISubjectProps } from '../SubjectTile';

describe('SubjectTile component', () => {
    /** Define required props */
    const requiredProps: ISubjectProps = {
        subject: 'foo',
    };

    test('Is match snapshot', () => {
        /** Render via enzyme */
        const root = shallow(<Component {...requiredProps} />);

        /** Assert component matches snapshot */
        expect(root).toMatchSnapshot();
    });
});
