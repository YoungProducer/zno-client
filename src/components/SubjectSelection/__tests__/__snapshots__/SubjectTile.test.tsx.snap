// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubjectTile component Is match snapshot 1`] = `
<WithStyles(ForwardRef(Grid))
  item={true}
  xs={3}
>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-button-1"
  >
    foo
  </WithStyles(ForwardRef(Button))>
</WithStyles(ForwardRef(Grid))>
`;
